src/stdlib_sorting_ord_sort.fypp

```diff
@@ -0,0 +1,1419 @@
+!!   option. This file may not be copied, modified, or distributed
+!!   except according to those terms.
+!!
+!! so the license for the original`slice.rs` code is compatible with the use
+!! of modified versions of the code in the Fortran Standard Library under
+!! the MIT license.
+
+submodule(stdlib_sorting) stdlib_sorting_ord_sort
+
+    implicit none
+
+contains
+
+#:for k1 in INT_KINDS
+
+    module subroutine ${k1}$_ord_sort( array, work )
+! A translation to Fortran 2008, of the `"Rust" sort` algorithm found in
+! `slice.rs`
+! https://github.com/rust-lang/rust/blob/90eb44a5897c39e3dff9c7e48e3973671dcd9496/src/liballoc/slice.rs#L2159
+! The Rust version in turn is a simplification of the Timsort algorithm
+! described in
+! https://svn.python.org/projects/python/trunk/Objects/listsort.txt, as
+! it drops both the use of 'galloping' to identify bounds of regions to be
+! sorted and the estimation of the optimal `run size`. However it remains
+! a hybrid sorting algorithm combining an iterative Merge sort controlled
+! by a stack of `RUNS` identified by regions of uniformly decreasing or
+! non-decreasing sequences that may be expanded to a minimum run size and
+! initially processed by an insertion sort.
+!
+! Note the Fortran implementation simplifies the logic as it only has to
+! deal with Fortran arrays of intrinsic types and not the full generality
+! of Rust's arrays and lists for arbitrary types. It also adds the
+! estimation of the optimal `run size` as suggested in Tim Peters'
+! original `listsort.txt`, and an optional `work` array to be used as
+! scratch memory.
+        integer(${k1}$), intent(inout)           :: array(0:)
+        integer(${k1}$), intent(inout), optional :: work(0:)
+
+        integer(${k1}$), allocatable :: buf(:)
+        integer(int_size) :: array_size
+        integer :: stat
+
+        if ( present(work) ) then
+! Use the work array as scratch memory
+            call merge_sort( array, work )
+        else
+! Allocate a buffer to use as scratch memory.
+            array_size = size( array, kind=int_size )
+            allocate( buf(0:array_size/2-1), stat=stat )
+            if ( stat /= 0 ) error stop "Allocation of buffer failed."
+            call merge_sort( array, buf )
+        end if
+
+    contains
+
+        pure function calc_min_run( n ) result(min_run)
+!! Returns the minimum length of a run from 32-63 so that N/MIN_RUN is
+!! less than or equal to a power of two. See
+!! https://svn.python.org/projects/python/trunk/Objects/listsort.txt
+            integer(int_size)             :: min_run
+            integer(int_size), intent(in) :: n
+
+            integer(int_size) :: num, r
+
+            num = n
+            r = 0_int_size
+
+            do while( num >= 64 )
+                r = ior( r, iand(num, 1_int_size) )
+                num = ishft(num, -1_int_size)
+            end do
+            min_run = num + r
+
+        end function calc_min_run
```

Comment:
The function `calc_min_run` appears multiple times, while it seems to be always the same. What is the reason to replicate it multiple times? Could it be kept outside the scope of a procedure?

Reply:
I thought `calc_min_run` was more likely to be inlined if made an internal procedure, but as it is invoked only once at the start of the main procedure, inlining only has a very minor impact on performance.

Reply:
If it is in the same module, I would expect that the compiler would inline it (if possile of course), even if it is not an internal procedure. I will leave this comment open for now.

Reply:
Not relevant anymore due to addition for Fypp pre-processing for generating code for integer, real, and string_type